peer/x/roundrobin/list.go

```diff
@@ -119,13 +102,26 @@ func (pl *List) Update(updates peer.ListUpdates) error {
 	defer pl.lock.Unlock()
 
 	var errs error
+	if pl.shouldRetainPeers.Load() {
+		for _, peerID := range removals {
+			errs = multierr.Append(errs, pl.removePeerIdentifier(peerID))
+		}
 
-	for _, peerID := range removals {
-		errs = multierr.Append(errs, pl.removePeerIdentifier(peerID))
+		for _, peerID := range additions {
+			errs = multierr.Append(errs, pl.addPeerIdentifier(peerID))
+		}
+		return errs
 	}
 
+	for _, peerID := range removals {
+		if _, ok := pl.uninitializedPeers[peerID.Identifier()]; ok {
```

Comment:
This function is a bit dense with its two paths. Consider making two helper functions:

```go
if pl.shouldRetainPeers.Load() {
  return pl.updateInitialized(updates)
}
return pl.updateUninitialized(updates)
```